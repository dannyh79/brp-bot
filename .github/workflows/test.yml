name: Test

on:
  push:
    branches: [main]
    paths-ignore:
      - 'docs/**'
      - 'README.md'
  pull_request:
    branches: [main]
    paths-ignore:
      - 'docs/**'
      - 'README.md'

env:
  NODE_VERSION: 22.x
  PNPM_VERSION: 10

jobs:
  typecheck:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: pnpm/action-setup@v4
        name: Install pnpm
        with:
          version: ${{ env.PNPM_VERSION }}
          run_install: false
      - name: Use Node.js ${{ env.NODE_VERSION }}
        uses: actions/setup-node@v4
        with:
          node-version: ${{ env.NODE_VERSION }}
          cache: pnpm
      - run: pnpm install --frozen-lockfile
      - run: pnpm tsc --noEmit

  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: pnpm/action-setup@v4
        name: Install pnpm
        with:
          version: ${{ env.PNPM_VERSION }}
          run_install: false
      - name: Use Node.js ${{ env.NODE_VERSION }}
        uses: actions/setup-node@v4
        with:
          node-version: ${{ env.NODE_VERSION }}
          cache: pnpm
      - run: pnpm install --frozen-lockfile
      - run: pnpm eslint .

  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: pnpm/action-setup@v4
        name: Install pnpm
        with:
          version: ${{ env.PNPM_VERSION }}
          run_install: false
      - name: Use Node.js ${{ env.NODE_VERSION }}
        uses: actions/setup-node@v4
        with:
          node-version: ${{ env.NODE_VERSION }}
          cache: pnpm
      - name: Copy wrangler config from template
        run: cp wrangler.toml.example wrangler.toml
      - run: pnpm install --frozen-lockfile
      - run: pnpm vitest --run --passWithNoTests --coverage
