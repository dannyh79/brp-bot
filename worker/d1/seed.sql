INSERT INTO plans (date, praise_scope, praise_content, devotional_scope) VALUES
  ('2025-01-01', '歷代志上 16:34 CCB', '你們要稱謝耶和華，因為祂是美善的，祂的慈愛永遠長存！', '出埃及記 第 8 章')
  ON CONFLICT (date) DO UPDATE SET
    praise_scope = '歷代志上 16:34 CCB',
    praise_content = '你們要稱謝耶和華，因為祂是美善的，祂的慈愛永遠長存！',
    devotional_scope = '出埃及記 第 8 章';

INSERT INTO recipients (id) VALUES
  ('YOUR-LINE-GROUP-ID')
  ON CONFLICT (id) DO NOTHING;
