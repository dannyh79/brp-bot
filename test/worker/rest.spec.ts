import { SELF, env } from 'cloudflare:test';
import * as helper from 'test/helpers/d1';

const stubDomain = 'https://brp-bot.pages.dev';

describe('GET /api/v1/plan', () => {
  beforeEach(async () => {
    await helper.insertPlanRecord();
  });

  it('responds 200 with plan for the date', async () => {
    const response = await SELF.fetch(`${stubDomain}/api/v1/plan?date=2025-01-01`);
    expect(response.status).toBe(200);
    expect(await response.json()).toMatchObject({
      date: '2025-01-01',
      praise: {
        scope: '歷代志上 16:34 CCB',
        content: '你們要稱謝耶和華，因為祂是美善的，祂的慈愛永遠長存！',
      },
      repentence:
        '上帝啊，求你按你的慈愛恩待我！\n按你極大的憐憫除去我 ___ 的過犯！\n求祢洗淨我的罪過，清除我的罪惡。\n求祢讓我重新享受蒙祢拯救的喜樂，賜我一個樂意順服祢的心靈，並開始以 ___ 的行動做出改變。',
      devotional: {
        scope: '出埃及記 第 8 章',
        content: ['1. 你覺得神透過今天的經文對你說什麼呢？', '2. 有什麼你可以做出的行動或改變呢？'],
      },
      prayer:
        '神啊！我將我的 ___ ， ___ ， ___ 交給祢，我相信祢在這些事上掌權。\n我們在天上的父：願人都尊你的名為聖。\n願你的國降臨；願你的旨意行在地上，如同行在天上。\n我們日用的飲食，今日賜給我們。\n免我們的債，如同我們免了人的債。\n不叫我們陷入試探；救我們脫離那惡者。\n因為國度、權柄、榮耀，全是你的，直到永遠。阿們！',
    });
  });

  it('responds 200 with plan in HTML', async () => {
    const response = await SELF.fetch(`${stubDomain}/api/v1/plan?date=2025-01-01&format=html`);
    expect(response.status).toBe(200);
    expect(await response.text()).toMatchSnapshot();
  });

  it('responds 404', async () => {
    const response = await SELF.fetch(`${stubDomain}/plan?date=2024-12-31`);
    expect(response.status).toBe(404);
  });
});

describe('POST /api/v1/recipients', () => {
  const recipient = helper.recipientRecordFixture;

  beforeEach(async () => {
    await helper.insertRecipientRecord();
  });

  it('responds 401 when not authorized', async () => {
    const response = await SELF.fetch(`${stubDomain}/api/v1/recipients`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ id: 'C5678f49365c6b492b337189e3343a9d9' }),
    });
    expect(response.status).toBe(401);
  });

  it('responds 204 when saves a recipient', async () => {
    const response = await SELF.fetch(`${stubDomain}/api/v1/recipients`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${env.API_TOKEN}` },
      body: JSON.stringify({ id: 'C5678f49365c6b492b337189e3343a9d9' }),
    });
    expect(response.status).toBe(204);
    expect(await response.text()).toBe('');
  });

  it('responds 304', async () => {
    const response = await SELF.fetch(`${stubDomain}/api/v1/recipients`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${env.API_TOKEN}` },
      body: JSON.stringify({ id: recipient.id }),
    });
    expect(response.status).toBe(304);
    expect(await response.text()).toBe('');
  });
});

describe('DELETE /api/v1/recipients/:id', () => {
  const recipient = helper.recipientRecordFixture;

  beforeEach(async () => {
    await helper.insertRecipientRecord();
  });

  it('responds 401 when not authorized', async () => {
    const response = await SELF.fetch(`${stubDomain}/api/v1/recipients/${recipient.id}`, {
      method: 'DELETE',
    });
    expect(response.status).toBe(401);
  });

  it('responds 204 when deletes a recipient', async () => {
    const response = await SELF.fetch(`${stubDomain}/api/v1/recipients/${recipient.id}`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${env.API_TOKEN}` },
    });
    expect(response.status).toBe(204);
    expect(await response.text()).toBe('');
  });

  it('responds 404 when recipient not found', async () => {
    const response = await SELF.fetch(
      `${stubDomain}/api/v1/recipients/C5678f49365c6b492b337189e3343a9d9`,
      {
        method: 'DELETE',
        headers: { Authorization: `Bearer ${env.API_TOKEN}` },
      },
    );
    expect(response.status).toBe(404);
    expect(await response.text()).toBe('');
  });
});
