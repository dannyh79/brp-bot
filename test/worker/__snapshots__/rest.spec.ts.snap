// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /api/v1/plan > responds 200 with plan in HTML, with querystring format=html 1`] = `
"<!DOCTYPE html>
    <html lang="zh-Hant">
      <head>
        <title>好好靈修 Daily Devotion - 2025-01-01</title>
        <meta
          name="description"
          content="[非官方] The Hope 好好靈修每日靈修內容。 PDF 版請至 thehope.co/brp 。"
        />
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <script defer src="https://cdn.jsdelivr.net/npm/@tailwindcss/browser@4"></script>
      </head>
      <body>
        <main class="max-w-md mx-2 md:mx-auto p-4 bg-[#FEFEFE] rounded-xl shadow-md text-[#5D5D5D]"><h1 class="bg-[#FFCC32] font-bold text-[#1D292E] text-2xl pl-4 py-4 rounded-lg mb-3">好好靈修 Daily Devotion</h1><article class="space-y-6 text-md leading-8"><div class="bg-[#EEF0F4] rounded-lg px-4 py-4"><h2 class="text-lg mb-4">1/1 星期三</h2><section class="mb-6 flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">讚美</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="pb-2 pl-6 w-full font-bold"><p>你們要稱謝耶和華，因為祂是美善的，祂的慈愛永遠長存！</p><p>歷代志上 16:34 CCB</p></div></section><section class="mb-6 flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">悔改</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="w-full space-y-2"><div class="rounded-2xl bg-[#1D292E] text-white ml-2 px-4 py-2"><p>聖靈懇求祢在今天光照我在生命中，有沒有什麼是祢要我去做，但我沒有去做的事？</p><p>又有什麼是祢不喜悅我去做，但我卻行的事？</p></div><div class="pb-2 pl-6"><p>上帝啊，求你按你的慈愛恩待我！</p><p>按你極大的憐憫除去我 ___ 的過犯！</p><p>求祢洗淨我的罪過，清除我的罪惡。</p><p>求祢讓我重新享受蒙祢拯救的喜樂，賜我一個樂意順服祢的心靈，並開始以 ___ 的行動做出改變。</p></div></div></section><section class="mb-6 flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">反思</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="w-full space-y-2"><div class="pl-2"><p class="font-bold text-xl">出埃及記 第 8 章</p><a href="https://www.bible.com/bible/1392/EXO.8" class="text-blue-500 hover:underline text-md block my-2" target="_blank" rel="noopener noreferrer">YouVersion 連結</a></div><div class="rounded-2xl bg-[#1D292E] text-white ml-2 mb-2 px-4 py-2"><p>1. 你覺得神透過今天的經文對你說什麼呢？</p><p>2. 有什麼你可以做出的行動或改變呢？</p></div></div></section><section class="flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">祈禱</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="w-full space-y-2"><p class="rounded-2xl bg-[#1D292E] text-white ml-2 px-4 py-2">神啊！我將我的 ___ ， ___ ， ___ 交給祢，我相信祢在這些事上掌權。</p><div class="pb-2 pl-6"><p>我們在天上的父：願人都尊你的名為聖。</p><p>願你的國降臨；願你的旨意行在地上，如同行在天上。</p><p>我們日用的飲食，今日賜給我們。</p><p>免我們的債，如同我們免了人的債。</p><p>不叫我們陷入試探；救我們脫離那惡者。</p><p>因為國度、權柄、榮耀，全是你的，直到永遠。阿們！</p></div></div></section></div></article></main>
      </body>
    </html>"
`;

exports[`GET /api/v1/plan > without date params > responds 200 with plan in HTML, with querystring format=html 1`] = `
"<!DOCTYPE html>
    <html lang="zh-Hant">
      <head>
        <title>好好靈修 Daily Devotion - 2025-01-01</title>
        <meta
          name="description"
          content="[非官方] The Hope 好好靈修每日靈修內容。 PDF 版請至 thehope.co/brp 。"
        />
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <script defer src="https://cdn.jsdelivr.net/npm/@tailwindcss/browser@4"></script>
      </head>
      <body>
        <main class="max-w-md mx-2 md:mx-auto p-4 bg-[#FEFEFE] rounded-xl shadow-md text-[#5D5D5D]"><h1 class="bg-[#FFCC32] font-bold text-[#1D292E] text-2xl pl-4 py-4 rounded-lg mb-3">好好靈修 Daily Devotion</h1><article class="space-y-6 text-md leading-8"><div class="bg-[#EEF0F4] rounded-lg px-4 py-4"><h2 class="text-lg mb-4">1/1 星期三</h2><section class="mb-6 flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">讚美</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="pb-2 pl-6 w-full font-bold"><p>你們要稱謝耶和華，因為祂是美善的，祂的慈愛永遠長存！</p><p>歷代志上 16:34 CCB</p></div></section><section class="mb-6 flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">悔改</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="w-full space-y-2"><div class="rounded-2xl bg-[#1D292E] text-white ml-2 px-4 py-2"><p>聖靈懇求祢在今天光照我在生命中，有沒有什麼是祢要我去做，但我沒有去做的事？</p><p>又有什麼是祢不喜悅我去做，但我卻行的事？</p></div><div class="pb-2 pl-6"><p>上帝啊，求你按你的慈愛恩待我！</p><p>按你極大的憐憫除去我 ___ 的過犯！</p><p>求祢洗淨我的罪過，清除我的罪惡。</p><p>求祢讓我重新享受蒙祢拯救的喜樂，賜我一個樂意順服祢的心靈，並開始以 ___ 的行動做出改變。</p></div></div></section><section class="mb-6 flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">反思</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="w-full space-y-2"><div class="pl-2"><p class="font-bold text-xl">出埃及記 第 8 章</p><a href="https://www.bible.com/bible/1392/EXO.8" class="text-blue-500 hover:underline text-md block my-2" target="_blank" rel="noopener noreferrer">YouVersion 連結</a></div><div class="rounded-2xl bg-[#1D292E] text-white ml-2 mb-2 px-4 py-2"><p>1. 你覺得神透過今天的經文對你說什麼呢？</p><p>2. 有什麼你可以做出的行動或改變呢？</p></div></div></section><section class="flex"><h3 class="font-bold text-[#1D292E] text-xl text-center">祈禱</h3><div class="flex flex-col items-center space-y-1"><span class="flex-none rounded-full bg-[#1D292E] w-3 h-3"></span><span class="rounded-lg bg-[#1D292E] w-0.75 h-full"></span></div><div class="w-full space-y-2"><p class="rounded-2xl bg-[#1D292E] text-white ml-2 px-4 py-2">神啊！我將我的 ___ ， ___ ， ___ 交給祢，我相信祢在這些事上掌權。</p><div class="pb-2 pl-6"><p>我們在天上的父：願人都尊你的名為聖。</p><p>願你的國降臨；願你的旨意行在地上，如同行在天上。</p><p>我們日用的飲食，今日賜給我們。</p><p>免我們的債，如同我們免了人的債。</p><p>不叫我們陷入試探；救我們脫離那惡者。</p><p>因為國度、權柄、榮耀，全是你的，直到永遠。阿們！</p></div></div></section></div></article></main>
      </body>
    </html>"
`;
